File: Integer Set/integer_set.rb
class Set
  def initialize
    @store = []
  end
  
  # O(n)
  def insert(val)
    @store.push(val) unless include?(val)
  end

  # O(n)
  def remove(val)
    @store.delete(val) if include?(val)
  end
  
  # O(n)
  def include?(val)
    @store.any? { |element| element == val }
  end
end

class MaxIntSet
  def initialize(max)
    @max = max
    @store = (0...max).map { |index| false }
  end

  # O(1)
  def insert(num)
    validate!(num)
    @store[num] = true
  end

  # O(1)
  def remove(num)
    validate!(num)
    @store[num] = false
  end

  # O(1)
  def include?(num)
    @store[num]
  end

  private
  def is_valid?(num)
    (num >= 0) && (num < @max)
  end

  def validate!(num)
    raise 'Out of bounds' unless is_valid?(num)
  end
end


class IntSet
  def initialize(num_buckets = 20)
    @num_buckets = num_buckets
    @store = Array.new(num_buckets) { Array.new }
  end

  # O(1)
  def insert(num)
    bucket = self[num]
    bucket.push(num)
  end

  # O(k)
  def remove(num)
    bucket = self[num]
    bucket.delete(num)
  end

  # O(k)
  def include?(num)
    bucket = self[num]
    bucket.include?(num)
  end

  private
  def [](num)
    @store[num % num_buckets]
  end

  def num_buckets
    @num_buckets
  end
end

class ResizingIntSet
  attr_reader :count

  def initialize(num_buckets = 20)
    @count = 0
    @store = Array.new(num_buckets) { Array.new }
  end

  # O(1) ammortized
  def insert(num)
    resize! if @count == num_buckets
    bucket = self[num]
    bucket.push(num)
    @count += 1
  end
  
  # O(1) but O(k) worst case
  def remove(num)
    bucket = self[num]
    bucket.delete(num)
    @count -= 1
  end
  
  # O(1) but O(k) worst case
  def include?(num)
    bucket = self[num]
    bucket.include?(num)
  end

  private
  def [](num)
    @store[num % num_buckets]
  end

  def num_buckets
    @store.length
  end

  # O(n) but O(nk) worst case
  def resize!
    new_num_buckets = num_buckets * 2
    new_store = Array.new(new_num_buckets) { Array.new }
    @store.each do |bucket|
      bucket.each do |element|
        new_bucket = new_store[element % new_num_buckets]
        new_bucket.push(element)
      end
    end
    @store = new_store
  end
end

File: Integer Set/IntegerSet.cpp
#include <string>
#include <vector>
#include <sstream>
#include <iostream>
#include "IntegerSet.hpp"

/* constructor */
IntegerSet::IntegerSet() : count{0}, numBuckets{5} {
  store = new Bucket[numBuckets];
  fill();
}

/* basic operations */
int IntegerSet::length() const {
  return count;
}

// O(1) ammortized
void IntegerSet::insert(int val) {
  if (count == numBuckets) resize();
  int internal = bucketIndex(val, numBuckets);
  Bucket* bucket = &store[internal];
  bucket->push(val);
  count++;
}

// O(1) but worst case O(k)
void IntegerSet::remove(int val) {
  Bucket* bucket = &store[bucketIndex(val, numBuckets)];
  bucket->remove(val);
  count--;
}

// O(1) but worst case O(k)
bool IntegerSet::includes(int val) const {
  int internal = bucketIndex(val, numBuckets);
  Bucket* bucket = &store[internal];
  return bucket->includes(val);
}

/* debugger */
void IntegerSet::print() const {
  std::cout << "{\n";
  for (int indexRow = 0; indexRow < numBuckets; indexRow++) {
    std::cout << "{ ";
    for (int indexCol = 0; indexCol < store[indexRow].length(); indexCol++) {
      std::cout << store[indexRow][indexCol];
      if (indexCol < store[indexRow].length() - 1) std::cout << ", ";
    }
    std::cout << " }";
    if (indexRow < numBuckets - 1) std::cout << ",\n";
  }
  std::cout << "\n}\n" << std::endl;
}

/* private */
void IntegerSet::fill() {
  for (int indexRow = 0; indexRow < numBuckets; indexRow++) {
    store[indexRow] = *(new Bucket);
  }
}

int IntegerSet::bucketIndex(int val, int nBuckets) const {
  int hashValue = hash(val);
  if (hashValue > 0) {
    return hashValue % nBuckets;
  } else {
    return (-1 * hashValue) % nBuckets;
  }
}

int IntegerSet::hash(int val) const {
  std::hash<std::string> hasher;
  std::stringstream string;
  string << val;
  return (int) hasher(string.str());
}

void IntegerSet::resize() {
  int value;
  int bucketIdx;
  Bucket* bucket;
  int new_buckets = numBuckets * 2;
  Bucket* new_store = new Bucket[new_buckets];
  
  for (int indexRow = 0; indexRow < new_buckets; indexRow++) {
    new_store[indexRow] = *(new Bucket);
  }
  
  for (int indexRow = 0; indexRow < numBuckets; indexRow++) {
    bucket = &store[indexRow];
    for (int indexCol = 0; indexCol < bucket->length(); indexCol++) {
      value = (*bucket)[indexCol];
      bucketIdx = bucketIndex(value, new_buckets);
      bucket = &new_store[bucketIdx];
      bucket->push(value);
    }
  }
  
  delete[] store;
  store = new_store;
  numBuckets = new_buckets;
}
